File: binding/src/printsupport/libqprintdialog.cpp
#include <QAbstractPrintDialog>
#include <QObject>
#include <QPrintDialog>
#include <QPrinter>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <type_traits>
#include <QWidget>
#include <qprintdialog.h>
#include "libqprintdialog.h"
#include "libqprintdialog.hxx"

QPrintDialog* QPrintDialog_new(QWidget* parent) {
	 return new VirtualQPrintDialog(parent);
}

QPrintDialog* QPrintDialog_new2(QPrinter* printer) {
	 return new VirtualQPrintDialog(printer);
}

QPrintDialog* QPrintDialog_new3() {
	 return new VirtualQPrintDialog();
}

QPrintDialog* QPrintDialog_new4(QPrinter* printer, QWidget* parent) {
	 return new VirtualQPrintDialog(printer, parent);
}

libqt_string QPrintDialog_Tr(const char* s) {
	QString _ret = QPrintDialog::tr(s);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

int QPrintDialog_Exec(QPrintDialog* self) {
	return self->exec();
}

void QPrintDialog_Done(QPrintDialog* self, int result) {
	self->done(result);
}

void QPrintDialog_SetOption(QPrintDialog* self, int option) {
	self->setOption(static_cast<QAbstractPrintDialog::PrintDialogOption>(option));
}

bool QPrintDialog_TestOption(const QPrintDialog* self, int option) {
	return self->testOption(static_cast<QAbstractPrintDialog::PrintDialogOption>(option));
}

void QPrintDialog_SetOptions(QPrintDialog* self, int options) {
	self->setOptions(static_cast<QFlags<QAbstractPrintDialog::PrintDialogOption>>(options));
}

int QPrintDialog_Options(const QPrintDialog* self) {
	return self->options();
}

void QPrintDialog_SetVisible(QPrintDialog* self, bool visible) {
	self->setVisible(visible);
}

void QPrintDialog_Open(QPrintDialog* self, QObject* receiver, const char* member) {
	self->open(receiver, member);
}

void QPrintDialog_Accepted(QPrintDialog* self, QPrinter* printer) {
	self->accepted(printer);
}

libqt_string QPrintDialog_Tr2(const char* s, const char* c) {
	QString _ret = QPrintDialog::tr(s, c);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QPrintDialog_Tr3(const char* s, const char* c, int n) {
	QString _ret = QPrintDialog::tr(s, c, n);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QPrintDialog_SetOption2(QPrintDialog* self, int option, bool on) {
	self->setOption(static_cast<QAbstractPrintDialog::PrintDialogOption>(option), on);
}

// Base class handler implementation
int QPrintDialog_QBaseExec(QPrintDialog* self) {
	auto* vqprintdialog = dynamic_cast<VirtualQPrintDialog*>(self);
	if (vqprintdialog && vqprintdialog->isVirtualQPrintDialog) {
vqprintdialog->setQPrintDialog_Exec_IsBase(true);
	return vqprintdialog->exec();
}
}

// Auxiliary method to allow providing re-implementation
void QPrintDialog_OnExec(QPrintDialog* self, intptr_t slot) {
	auto* vqprintdialog = dynamic_cast<VirtualQPrintDialog*>(self);
	if (vqprintdialog && vqprintdialog->isVirtualQPrintDialog) {
vqprintdialog->setQPrintDialog_Exec_Callback(reinterpret_cast<VirtualQPrintDialog::QPrintDialog_Exec_Callback>(slot));
}
}

// Base class handler implementation
void QPrintDialog_QBaseDone(QPrintDialog* self, int result) {
	auto* vqprintdialog = dynamic_cast<VirtualQPrintDialog*>(self);
	if (vqprintdialog && vqprintdialog->isVirtualQPrintDialog) {
vqprintdialog->setQPrintDialog_Done_IsBase(true);
	vqprintdialog->done(result);
}
}

// Auxiliary method to allow providing re-implementation
void QPrintDialog_OnDone(QPrintDialog* self, intptr_t slot) {
	auto* vqprintdialog = dynamic_cast<VirtualQPrintDialog*>(self);
	if (vqprintdialog && vqprintdialog->isVirtualQPrintDialog) {
vqprintdialog->setQPrintDialog_Done_Callback(reinterpret_cast<VirtualQPrintDialog::QPrintDialog_Done_Callback>(slot));
}
}

// Base class handler implementation
void QPrintDialog_QBaseSetVisible(QPrintDialog* self, bool visible) {
	auto* vqprintdialog = dynamic_cast<VirtualQPrintDialog*>(self);
	if (vqprintdialog && vqprintdialog->isVirtualQPrintDialog) {
vqprintdialog->setQPrintDialog_SetVisible_IsBase(true);
	vqprintdialog->setVisible(visible);
}
}

// Auxiliary method to allow providing re-implementation
void QPrintDialog_OnSetVisible(QPrintDialog* self, intptr_t slot) {
	auto* vqprintdialog = dynamic_cast<VirtualQPrintDialog*>(self);
	if (vqprintdialog && vqprintdialog->isVirtualQPrintDialog) {
vqprintdialog->setQPrintDialog_SetVisible_Callback(reinterpret_cast<VirtualQPrintDialog::QPrintDialog_SetVisible_Callback>(slot));
}
}

void QPrintDialog_Delete(QPrintDialog* self) {
    delete self;
}

File: binding/include/printsupport/libqprintdialog.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qprintdialog.html
pub const qprintdialog = struct {

    /// New constructs a new QPrintDialog object.
    pub fn New(parent: ?*anyopaque) QtC.QPrintDialog {
        return qtc.QPrintDialog_new(@ptrCast(parent));
    }


    /// New2 constructs a new QPrintDialog object.
    pub fn New2(printer: ?*anyopaque) QtC.QPrintDialog {
        return qtc.QPrintDialog_new2(@ptrCast(printer));
    }


    /// New3 constructs a new QPrintDialog object.
    pub fn New3() QtC.QPrintDialog {
        return qtc.QPrintDialog_new3();
    }


    /// New4 constructs a new QPrintDialog object.
    pub fn New4(printer: ?*anyopaque, parent: ?*anyopaque) QtC.QPrintDialog {
        return qtc.QPrintDialog_new4(@ptrCast(printer), @ptrCast(parent));
    }


    pub fn Tr(s: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QPrintDialog_Tr(@ptrCast(s));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qprintdialog.Tr: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Exec(self: ?*anyopaque, ) i32 {
        return qtc.QPrintDialog_Exec(@ptrCast(self));
    }

    /// Allows for overriding the related default method
    pub fn OnExec(self: ?*anyopaque, callback: *const fn (?*anyopaque) callconv(.c) i32) void {
        qtc.QPrintDialog_OnExec(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseExec(self: ?*anyopaque, ) i32 {
        return qtc.QPrintDialog_QBaseExec(@ptrCast(self));
    }

    pub fn Done(self: ?*anyopaque, result: i32) void {
        qtc.QPrintDialog_Done(@ptrCast(self), result);
    }

    /// Allows for overriding the related default method
    pub fn OnDone(self: ?*anyopaque, callback: *const fn (?*anyopaque, i32) callconv(.c) void) void {
        qtc.QPrintDialog_OnDone(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseDone(self: ?*anyopaque, result: i32) void {
        qtc.QPrintDialog_QBaseDone(@ptrCast(self), result);
    }

    pub fn SetOption(self: ?*anyopaque, option: i32) void {
        qtc.QPrintDialog_SetOption(@ptrCast(self), @intCast(option));
    }

    pub fn TestOption(self: ?*anyopaque, option: i32) bool {
        return qtc.QPrintDialog_TestOption(@ptrCast(self), @intCast(option));
    }

    pub fn SetOptions(self: ?*anyopaque, options: i32) void {
        qtc.QPrintDialog_SetOptions(@ptrCast(self), @intCast(options));
    }

    pub fn Options(self: ?*anyopaque, ) i32 {
        return qtc.QPrintDialog_Options(@ptrCast(self));
    }

    pub fn SetVisible(self: ?*anyopaque, visible: bool) void {
        qtc.QPrintDialog_SetVisible(@ptrCast(self), visible);
    }

    /// Allows for overriding the related default method
    pub fn OnSetVisible(self: ?*anyopaque, callback: *const fn (?*anyopaque, bool) callconv(.c) void) void {
        qtc.QPrintDialog_OnSetVisible(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseSetVisible(self: ?*anyopaque, visible: bool) void {
        qtc.QPrintDialog_QBaseSetVisible(@ptrCast(self), visible);
    }

    pub fn Open(self: ?*anyopaque, receiver: ?*anyopaque, member: []const u8) void {
        qtc.QPrintDialog_Open(@ptrCast(self), @ptrCast(receiver), @ptrCast(member));
    }

    pub fn Accepted(self: ?*anyopaque, printer: ?*anyopaque) void {
        qtc.QPrintDialog_Accepted(@ptrCast(self), @ptrCast(printer));
    }

    pub fn Tr2(s: []const u8, c: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QPrintDialog_Tr2(@ptrCast(s), @ptrCast(c));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qprintdialog.Tr2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Tr3(s: []const u8, c: []const u8, n: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QPrintDialog_Tr3(@ptrCast(s), @ptrCast(c), n);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qprintdialog.Tr3: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SetOption2(self: ?*anyopaque, option: i32, on: bool) void {
        qtc.QPrintDialog_SetOption2(@ptrCast(self), @intCast(option), on);
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QPrintDialog_Delete(@ptrCast(self));
    }
};
